File: src/tauri/secrets.ts
/**
 * Secrets Manager Tauri IPC 래퍼
 *
 * SecretManager를 통해 시크릿을 안전하게 관리합니다.
 * - Keychain 접근은 마스터키 로드 시 1회만 발생
 * - 나머지는 암호화된 Vault 파일에서 관리
 */

import { invoke, isTauriRuntime } from '@/tauri/invoke';

/**
 * 시크릿 엔트리 (key-value 쌍)
 */
export interface SecretEntry {
  key: string;
  value: string;
}

/**
 * 초기화 결과
 */
export interface SecretsInitResult {
  initialized: boolean;
  cached_count: number;
}

/**
 * 마이그레이션 결과
 */
export interface MigrationResult {
  migrated: number;
  failed: number;
  details: string[];
}

/**
 * SecretManager 초기화
 *
 * 앱 시작 시 1회 호출하여 마스터키를 로드하고 Vault를 복호화합니다.
 * 이 명령 호출 시 Keychain 프롬프트가 최대 1회 발생할 수 있습니다.
 */
export async function initializeSecrets(): Promise<SecretsInitResult> {
  if (!isTauriRuntime()) {
    console.warn('[Secrets] Not in Tauri runtime, skipping initialization');
    return { initialized: false, cached_count: 0 };
  }

  try {
    const result = await invoke<SecretsInitResult>('secrets_initialize');
    console.log('[Secrets] Initialized, cached secrets:', result.cached_count);
    return result;
  } catch (error) {
    console.error('[Secrets] Initialization failed:', error);
    throw error;
  }
}

/**
 * 시크릿 조회 (단일)
 *
 * Keychain 프롬프트 없이 메모리 캐시에서 조회합니다.
 */
export async function getSecret(key: string): Promise<string | null> {
  if (!isTauriRuntime()) {
    return null;
  }

  return await invoke<string | null>('secrets_get_one', { key });
}

/**
 * 시크릿 조회 (여러 개)
 *
 * Keychain 프롬프트 없이 메모리 캐시에서 조회합니다.
 */
export async function getSecrets(keys: string[]): Promise<Record<string, string>> {
  if (!isTauriRuntime()) {
    return {};
  }

  return await invoke<Record<string, string>>('secrets_get', { keys });
}

/**
 * 시크릿 저장 (단일)
 */
export async function setSecret(key: string, value: string): Promise<void> {
  if (!isTauriRuntime()) {
    console.warn('[Secrets] Not in Tauri runtime, skipping save');
    return;
  }

  await invoke<void>('secrets_set_one', { key, value });
}

/**
 * 시크릿 저장 (여러 개)
 */
export async function setSecrets(entries: SecretEntry[]): Promise<void> {
  if (!isTauriRuntime()) {
    console.warn('[Secrets] Not in Tauri runtime, skipping save');
    return;
  }

  await invoke<void>('secrets_set', { entries });
}

/**
 * 시크릿 삭제
 */
export async function deleteSecrets(keys: string[]): Promise<void> {
  if (!isTauriRuntime()) {
    return;
  }

  await invoke<void>('secrets_delete', { keys });
}

/**
 * 시크릿 존재 여부 확인
 *
 * Keychain 프롬프트 없이 확인합니다.
 */
export async function hasSecret(key: string): Promise<boolean> {
  if (!isTauriRuntime()) {
    return false;
  }

  return await invoke<boolean>('secrets_has', { key });
}

/**
 * 특정 prefix로 시작하는 시크릿 키 목록 조회
 */
export async function listSecretKeys(prefix: string): Promise<string[]> {
  if (!isTauriRuntime()) {
    return [];
  }

  return await invoke<string[]>('secrets_list_keys', { prefix });
}

/**
 * 기존 Keychain 엔트리를 Vault로 마이그레이션
 *
 * Settings에서 사용자가 명시적으로 실행하는 것을 권장합니다.
 * 마이그레이션 성공 시 기존 Keychain 엔트리는 삭제됩니다.
 */
export async function migrateLegacySecrets(): Promise<MigrationResult> {
  if (!isTauriRuntime()) {
    return { migrated: 0, failed: 0, details: [] };
  }

  return await invoke<MigrationResult>('secrets_migrate_legacy');
}

File: src-tauri/src/commands/secrets.rs
//! Secret Manager Tauri 명령어
//!
//! SecretManager를 통해 시크릿을 안전하게 관리합니다.
//! - 모든 시크릿은 메모리 캐시 + 암호화된 vault 파일에 저장
//! - Keychain 접근은 마스터키 로드 시 1회만 발생

use crate::error::{CommandError, CommandResult};
use crate::secrets::{MigrationResult, SECRETS};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// 시크릿 설정 요청
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretEntry {
    pub key: String,
    pub value: String,
}

/// 시크릿 초기화 결과
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretsInitResult {
    pub success: bool,
    pub cached_count: usize,
}

fn map_secret_error(err: crate::secrets::manager::SecretManagerError) -> CommandError {
    CommandError {
        code: "SECRET_MANAGER_ERROR".to_string(),
        message: format!("Secret manager error: {}", err),
        details: None,
    }
}

/// SecretManager 초기화
/// 
/// 앱 시작 시 호출하여 마스터키를 로드하고 vault를 복호화합니다.
/// 이 명령 호출 시 Keychain 프롬프트가 최대 1회 발생할 수 있습니다.
#[tauri::command]
pub async fn secrets_initialize() -> CommandResult<SecretsInitResult> {
    SECRETS
        .initialize()
        .await
        .map_err(map_secret_error)?;

    // 캐시된 시크릿 수 반환
    let count = SECRETS
        .list_keys_by_prefix("")
        .await
        .map_err(map_secret_error)?
        .len();

    Ok(SecretsInitResult {
        success: true,
        cached_count: count,
    })
}

/// 시크릿 조회
/// 
/// 여러 키를 한 번에 조회할 수 있습니다.
/// Keychain 프롬프트 없이 메모리 캐시에서 조회합니다.
#[tauri::command]
pub async fn secrets_get(keys: Vec<String>) -> CommandResult<HashMap<String, String>> {
    SECRETS
        .get_many(&keys)
        .await
        .map_err(map_secret_error)
}

/// 단일 시크릿 조회
#[tauri::command]
pub async fn secrets_get_one(key: String) -> CommandResult<Option<String>> {
    SECRETS
        .get(&key)
        .await
        .map_err(map_secret_error)
}

/// 시크릿 저장
/// 
/// 여러 키-값 쌍을 한 번에 저장할 수 있습니다.
/// 저장 후 vault 파일이 업데이트됩니다.
#[tauri::command]
pub async fn secrets_set(entries: Vec<SecretEntry>) -> CommandResult<()> {
    let entries: Vec<(String, String)> = entries
        .into_iter()
        .map(|e| (e.key, e.value))
        .collect();

    SECRETS
        .set_many(entries)
        .await
        .map_err(map_secret_error)
}

/// 단일 시크릿 저장
#[tauri::command]
pub async fn secrets_set_one(key: String, value: String) -> CommandResult<()> {
    SECRETS
        .set(&key, &value)
        .await
        .map_err(map_secret_error)
}

/// 시크릿 삭제
/// 
/// 여러 키를 한 번에 삭제할 수 있습니다.
#[tauri::command]
pub async fn secrets_delete(keys: Vec<String>) -> CommandResult<()> {
    SECRETS
        .delete_many(&keys)
        .await
        .map_err(map_secret_error)
}

/// 시크릿 존재 여부 확인
/// 
/// Keychain 프롬프트 없이 확인합니다.
#[tauri::command]
pub async fn secrets_has(key: String) -> CommandResult<bool> {
    SECRETS
        .has(&key)
        .await
        .map_err(map_secret_error)
}

/// 특정 prefix로 시작하는 모든 키 조회
#[tauri::command]
pub async fn secrets_list_keys(prefix: String) -> CommandResult<Vec<String>> {
    SECRETS
        .list_keys_by_prefix(&prefix)
        .await
        .map_err(map_secret_error)
}

/// 기존 Keychain 엔트리를 Vault로 마이그레이션
/// 
/// Settings에서 사용자가 명시적으로 호출합니다.
/// 마이그레이션 성공 시 기존 Keychain 엔트리는 삭제됩니다.
#[tauri::command]
pub async fn secrets_migrate_legacy() -> CommandResult<MigrationResult> {
    SECRETS
        .migrate_from_legacy_keychain()
        .await
        .map_err(map_secret_error)
}

